client/heartbeatstop_test.go

```diff
@@ -4,70 +4,136 @@
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/helper/pointer"
+	"github.com/hashicorp/nomad/helper/testlog"
 	"github.com/hashicorp/nomad/helper/uuid"
 	"github.com/hashicorp/nomad/nomad/structs"
-	"github.com/hashicorp/nomad/testutil"
 	"github.com/shoenig/test/must"
+	"github.com/shoenig/test/wait"
 )
 
-func TestHeartbeatStop_allocHook(t *testing.T) {
+func TestHeartbeatStop(t *testing.T) {
 	ci.Parallel(t)
 
-	server, _, cleanupS1 := testServer(t, nil)
-	defer cleanupS1()
-	testutil.WaitForLeader(t, server.RPC)
+	shutdownCh := make(chan struct{})
+	t.Cleanup(func() { close(shutdownCh) })
 
-	client, cleanupC1 := TestClient(t, func(c *config.Config) {
-		c.RPCHandler = server
-	})
-	defer cleanupC1()
+	destroyers := map[string]*mockAllocRunnerDestroyer{}
+
+	stopper := newHeartbeatStop(func(id string) (interfaces.AllocRunner, error) {
+		return destroyers[id], nil
+	},
+		time.Hour, // start grace, ignored in this test
+		testlog.HCLogger(t),
+		shutdownCh)
 
 	// an allocation, with a tiny lease
-	d := 1 * time.Microsecond
-	alloc := &structs.Allocation{
+	alloc1 := &structs.Allocation{
 		ID:        uuid.Generate(),
 		TaskGroup: "foo",
 		Job: &structs.Job{
 			TaskGroups: []*structs.TaskGroup{
 				{
 					Name: "foo",
 					Disconnect: &structs.DisconnectStrategy{
-						StopOnClientAfter: &d,
+						StopOnClientAfter: pointer.Of(time.Microsecond),
 					},
 				},
 			},
 		},
-		Resources: &structs.Resources{
-			CPU:      100,
-			MemoryMB: 100,
-			DiskMB:   0,
-		},
 	}
 
-	// alloc added to heartbeatStop.allocs
-	err := client.addAlloc(alloc, "")
-	must.NoError(t, err)
-	testutil.WaitForResult(func() (bool, error) {
-		client.heartbeatLock.Lock()
-		_, ok := client.heartbeatStop.allocInterval[alloc.ID]
-		client.heartbeatLock.Unlock()
-		return ok, nil
-	}, func(err error) {
-		must.NoError(t, err)
-	})
-
-	// the tiny lease causes the watch loop to destroy it
-	testutil.WaitForResult(func() (bool, error) {
-		_, ok := client.heartbeatStop.allocInterval[alloc.ID]
-		return !ok, nil
-	}, func(err error) {
-		must.NoError(t, err)
-	})
-
-	must.Nil(t, client.allocs[alloc.ID])
+	// an alloc with a longer lease
+	alloc2 := alloc1.Copy()
+	alloc2.ID = uuid.Generate()
+	alloc2.Job.TaskGroups[0].Disconnect.StopOnClientAfter = pointer.Of(2 * time.Second)
+
+	// an alloc with no disconnect config
+	alloc3 := alloc1.Copy()
+	alloc3.ID = uuid.Generate()
+	alloc3.Job.TaskGroups[0].Disconnect = nil
+
+	destroyers[alloc1.ID] = &mockAllocRunnerDestroyer{}
+	destroyers[alloc2.ID] = &mockAllocRunnerDestroyer{}
+	destroyers[alloc3.ID] = &mockAllocRunnerDestroyer{}
+
+	go stopper.watch()
+	stopper.allocHook(alloc1)
+	stopper.allocHook(alloc2)
+	stopper.allocHook(alloc3)
+
+	must.Wait(t, wait.InitialSuccess(
+		wait.Timeout(time.Second),
+		wait.ErrorFunc(func() error {
+			if destroyers[alloc1.ID].checkCalls() != 1 {
+				return fmt.Errorf("first alloc was not destroyed as expected")
```

Comment:
I know something in the past has linted `fmt.Errorf`s without any `%`s in em. I suppose there's no active check for that?  I think it would want all these to be

```suggestion
				return errors.New("first alloc was not destroyed as expected")
```